### PR TITLE
Issue #3215146 by Ressinel: As a LU I want to be able to mute and unmute Group notifications.

### PR DESCRIPTION
**Pull request by Ressinel**
## Problem
Being a member of a Group means you get a lot of notifications for actions happening in or around the Group. Posting on the stream, group content being created. This can be overwhelming.

## Solution
1. As a LU I should be able to mute/unmute Group notifications from the Group pages
2. As a LU whenever I mute a Group I don’t want to get any E-mail notifications about that Group
3. As a LU whenever I mute a Group I don’t want to get any notification center notifications about that Group.
4. As a LU whenever I mute a Group I don’t want to get any Push notifications about that Group.

## Issue tracker
- https://getopensocial.atlassian.net/browse/DS-7598
- https://getopensocial.atlassian.net/browse/DS-7599
- https://www.drupal.org/project/social/issues/3215146
- https://www.drupal.org/project/social/issues/3215148
- https://www.drupal.org/project/socialbase/issues/3216210
- https://www.drupal.org/project/socialblue/issues/3229814

## How to test
- [x] Implement socialbase theme [PR changes](https://git.drupalcode.org/project/socialbase/-/merge_requests/25)
```
"drupal/socialbase": {
    "Issue #3216210 by Ressinel: As a LU I want to be able to mute and unmute Group notifications.": "https://git.drupalcode.org/project/socialbase/-/merge_requests/25.diff"
}
```
- [x] Go to the group you joined
- [x] You will be able to mute/unmute group notifications including any mentions and welcome message (only if you are the group creator)
- [x] Muting the group notifications will result in their notifications not being sent to you anymore

## Screenshots
Mute group:

mute

Unmute group:

unmute

## Release notes
Added possibility to mute and unmute Group notifications. Muting the group notifications will result in their notifications not being send you anymore.

## Change Record
N/A

## Translations
N/A

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -140,7 +140,7 @@
         "drupal/exif_orientation": "^1.0",
         "drupal/features": "3.8",
         "drupal/field_group": "3.1",
-        "drupal/flag": "4.0-alpha3",
+        "drupal/flag": "4.0-beta2",
```
Comment (navneet0693) [inline composer.json:143]: Is this necessary for this feature? If yes, please add it the problem and solution section of the description.

File: modules/custom/activity_send/modules/activity_send_email/src/Plugin/QueueWorker/ActivitySendEmailWorker.php [hunk 2]
```diff
@@ -262,9 +275,31 @@ private function sendToFrequencyManager(array $parameters) {
 
     // We load all the target accounts.
     if ($target_accounts = $user_storage->loadMultiple($parameters['target_recipients'])) {
+      // Get the group from the activity.
+      $activity = $parameters['activity'];
```
Comment (navneet0693) [inline modules/custom/activity_send/modules/activity_send_email/src/Plugin/QueueWorker/ActivitySendEmailWorker.php:279]: Please refactor the code to use `$related_entity = $parameters['activity']->getRelatedEntity()` also, move this $related_entity outside if statement which can be then re-used.

File: modules/custom/activity_send/modules/activity_send_email/src/Plugin/QueueWorker/ActivitySendEmailWorker.php [hunk 3]
```diff
@@ -262,9 +275,31 @@ private function sendToFrequencyManager(array $parameters) {
...
+          $group = $related_entity->get('field_recipient_group')->first()->get('entity')->getTarget()->getValue();
+        }
+      }
+
       /** @var \Drupal\user\Entity\User $target_account */
       foreach ($target_accounts as $target_account) {
         if (($target_account instanceof User) && !$target_account->isBlocked()) {
+          // Need check if we have the group.
+          if (isset($group) && $group instanceof GroupInterface) {
+            // Only for users that didn't mute group notification.
```
Comment (navneet0693) [inline modules/custom/activity_send/modules/activity_send_email/src/Plugin/QueueWorker/ActivitySendEmailWorker.php:297]: Suggestion: Skip the notification for users which have muted the group notification in which this content was posted.

File: modules/custom/activity_send/modules/activity_send_email/src/Plugin/QueueWorker/ActivitySendEmailWorker.php [hunk 4]
```diff
@@ -262,9 +275,31 @@ private function sendToFrequencyManager(array $parameters) {
...
+        if (!$related_entity->get('field_recipient_group')->isEmpty()) {
+          /** @var \Drupal\group\Entity\GroupInterface $group */
+          $group = $related_entity->get('field_recipient_group')->first()->get('entity')->getTarget()->getValue();
+        }
+      }
+
       /** @var \Drupal\user\Entity\User $target_account */
       foreach ($target_accounts as $target_account) {
         if (($target_account instanceof User) && !$target_account->isBlocked()) {
+          // Need check if we have the group.
```
Comment (navneet0693) [inline modules/custom/activity_send/modules/activity_send_email/src/Plugin/QueueWorker/ActivitySendEmailWorker.php:295]: Suggestion: Check if we have $group set which means that this content was posted in a group.

File: modules/social_features/social_group/social_group.install [hunk 5]
```diff
@@ -640,3 +642,74 @@ function social_group_update_8906() {
...
   }
 }
+
+/**
+ * Helper function to import configurations.
+ *
+ * @param array $configs
+ *   Array of configuration file names and their folders.
+ */
+function _social_group_features_import_config_helper(array $configs) {
```
Comment (navneet0693) [inline modules/social_features/social_group/social_group.install:652]: Move this `social_core` module and make it a service class. Also, make this function a static function.

File: modules/social_features/social_group/social_group.install [hunk 6]
```diff
@@ -640,3 +642,74 @@ function social_group_update_8906() {
...
+        }
+      }
+    }
+  }
+}
+
+/**
+ * Import flag "Mute Group Notifications" configurations.
+ */
+function social_group_update_8907() {
```
Comment (navneet0693) [inline modules/social_features/social_group/social_group.install:673]: Start update hook numbers with 10201

File: modules/social_features/social_group/social_group.install [hunk 7]
```diff
@@ -640,3 +642,74 @@ function social_group_update_8906() {
...
+  ];
+
+  _social_group_features_import_config_helper($configs);
+}
+
+/**
+ * Update the permissions for social_group.
+ */
+function social_group_update_8908() {
+  _social_group_set_permissions();
```
Comment (navneet0693) [inline modules/social_features/social_group/social_group.install:687]: Instead just add the required permissions to required roles, we don't to accidentally change the current permission changes.

File: modules/social_features/social_group/social_group.install [hunk 8]
```diff
@@ -640,3 +642,74 @@ function social_group_update_8906() {
...
+
+/**
+ * Update User Groups view configurations.
+ */
+function social_group_update_8909() {
+  /** @var \Drupal\update_helper\Updater $updateHelper */
+  $updateHelper = \Drupal::service('update_helper.updater');
+
+  // Execute configuration update definitions with logging of success.
+  $updateHelper->executeUpdate('social_group', 'social_group_update_8908');
```
Comment (navneet0693) [inline modules/social_features/social_group/social_group.install:698]: Update hook numbers appended at the end of the files don't match.

File: modules/social_features/social_group/social_group.module [hunk 9]
```diff
@@ -2021,6 +2025,21 @@ function social_group_get_allowed_visibility_options_per_group_type($group_type_
  * Implements hook_views_query_alter().
  */
 function social_group_views_query_alter(ViewExecutable $view, QueryPluginBase $query) {
+  // User Groups view.
+  if ($view->id() == 'groups' && $view->getDisplay()->display['id'] == 'page_user_groups') {
+    // Change Join type if we want to see muted groups.
+    // @todo: Remove when will be fixed issue where exposed flag filter can't
```
Comment (navneet0693) [inline modules/social_features/social_group/social_group.module:2031]: If you have an existing issue filed or create an issue in the issue queue which explains the same and please add it here, or extend the documentation to explain the proper problem.

File: modules/social_features/social_group/config/install/flag.flag.mute_group_notify.yml [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+dependencies:
+  module:
+    - group
+id: mute_group_notify
+label: 'Mute Group Notify'
+bundles: {  }
+entity_type: group
+global: false
+weight: 0
+flag_short: 'Mute notify'
```
Comment (navneet0693) [inline modules/social_features/social_group/config/install/flag.flag.mute_group_notify.yml:12]: 'Mute'

File: modules/social_features/social_group/config/install/flag.flag.mute_group_notify.yml [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+flag_long: 'Muting the group notifications will result in their notifications not being send you anymore.'
+flag_message: ''
+unflag_short: 'Unmute notify'
```
Comment (navneet0693) [inline modules/social_features/social_group/config/install/flag.flag.mute_group_notify.yml:15]: 'Unmute'

File: modules/social_features/social_group/config/install/flag.flag.mute_group_notify.yml [hunk 12]
```diff
@@ -0,0 +1,30 @@
+langcode: en
+status: true
+dependencies:
+  module:
+    - group
+id: mute_group_notify
```
Comment (navneet0693) [inline modules/social_features/social_group/config/install/flag.flag.mute_group_notify.yml:6]: /s

`id: mute_group_notifications`

**Comment by navneet0693** [comment]
Screenshot 2021-06-15 at 5 49 17 PM
We need to theme the button a bit better.

Screenshot 2021-06-15 at 5 53 20 PM
The mentions notifications are still going through. In the above screenshot, Susan Williams had already muted the group notifications, but when Chris Hall posted something in the group mentioning Susan, Susan received the email. Is this intension behavior?

However, I do think that mentions email should be allowed to go.

**Comment by navneet0693** [comment]
@Ressinel Please also add screenshots mainly about the button being added as it is a visual change. Also, please enhance the release notes by adding more explanation.

**Comment by ronaldtebrake** [comment]
Hi @Ressinel @navneet0693 

What could be a nice addition is when a Group is muted, also the welcome messages are not shown anymore. To give the group managers a way to mute all those as well!

Not sure if that is too big of a scope creap, let me know, but this came in as feedback and thought it might make sense to add it to this scope already instead of adding it as a follow-up, i'd be happy to discuss the impact :D

**Comment by Ressinel** [comment]
> Screenshot 2021-06-15 at 5 53 20 PM
> The mentions notifications are still going through. In the above screenshot, Susan Williams had already muted the group notifications, but when Chris Hall posted something in the group mentioning Susan, Susan received the email. Is this intension behavior?
> 
> However, I do think that mentions email should be allowed to go.

Hi @ronaldtebrake 
What do you think about it?

File: composer.json (hunk 1 above)
Comment (Ressinel) [reply composer.json]: I forget to revert it back since we don't need it until we will use a select exposed filter for Mute/Unmute.
So I reverted the version back for now.

File: modules/custom/activity_send/modules/activity_send_email/src/Plugin/QueueWorker/ActivitySendEmailWorker.php (hunk 2 above)
Comment (Ressinel) [reply modules/custom/activity_send/modules/activity_send_email/src/Plugin/QueueWorker/ActivitySendEmailWorker.php]: Check how it looks now.

File: modules/custom/activity_send/modules/activity_send_email/src/Plugin/QueueWorker/ActivitySendEmailWorker.php (hunk 3 above)
Comment (Ressinel) [reply modules/custom/activity_send/modules/activity_send_email/src/Plugin/QueueWorker/ActivitySendEmailWorker.php]: Changed.

File: modules/custom/activity_send/modules/activity_send_email/src/Plugin/QueueWorker/ActivitySendEmailWorker.php (hunk 4 above)
Comment (Ressinel) [reply modules/custom/activity_send/modules/activity_send_email/src/Plugin/QueueWorker/ActivitySendEmailWorker.php]: Changed.

File: modules/social_features/social_group/social_group.install (hunk 5 above)
Comment (Ressinel) [reply modules/social_features/social_group/social_group.install]: Moved.

File: modules/social_features/social_group/social_group.install (hunk 6 above)
Comment (Ressinel) [reply modules/social_features/social_group/social_group.install]: Updated.

File: modules/social_features/social_group/social_group.install (hunk 7 above)
Comment (Ressinel) [reply modules/social_features/social_group/social_group.install]: Check how it looks now.

File: modules/social_features/social_group/social_group.install (hunk 8 above)
Comment (Ressinel) [reply modules/social_features/social_group/social_group.install]: Updated.

File: modules/social_features/social_group/social_group.module (hunk 9 above)
Comment (Ressinel) [reply modules/social_features/social_group/social_group.module:2031]: Check if it looks good to you now.

File: modules/social_features/social_group/config/install/flag.flag.mute_group_notify.yml (hunk 10 above)
Comment (Ressinel) [reply modules/social_features/social_group/config/install/flag.flag.mute_group_notify.yml]: Moritz suggested calls it a "Mute group". So I renamed it "Mute group".

File: modules/social_features/social_group/config/install/flag.flag.mute_group_notify.yml (hunk 11 above)
Comment (Ressinel) [reply modules/social_features/social_group/config/install/flag.flag.mute_group_notify.yml]: Moritz suggested calls it a "Unmute group". So I renamed it "Unmute group".

File: modules/social_features/social_group/config/install/flag.flag.mute_group_notify.yml (hunk 12 above)
Comment (Ressinel) [reply modules/social_features/social_group/config/install/flag.flag.mute_group_notify.yml]: Changed.

**Comment by Ressinel** [comment]
@navneet0693 

> @Ressinel Please also add screenshots mainly about the button being added as it is a visual change. Also, please enhance the release notes by adding more explanation.

Updated.

**Comment by Ressinel** [comment]
@navneet0693 

> Screenshot 2021-06-15 at 5 49 17 PM
> We need to theme the button a bit better.

Moritz suggested to place it in the drop down for now above the "leave group" option. So I move it there.

> Screenshot 2021-06-15 at 5 53 20 PM
> The mentions notifications are still going through. In the above screenshot, Susan Williams had already muted the group notifications, but when Chris Hall posted something in the group mentioning Susan, Susan received the email. Is this intension behavior?
> 
> However, I do think that mentions email should be allowed to go.

I muted these notifications as well.

**Comment by Ressinel** [comment]
@navneet0693 Also I added mute for Group Welcome Message in case of user mute Group notifications.

File: modules/custom/activity_basics/src/Plugin/ActivityContext/ContentInMyGroupActivityContext.php [hunk 13]
```diff
@@ -18,6 +23,60 @@
...
+  public function __construct(
+    array $configuration,
+    $plugin_id,
+    $plugin_definition,
+    QueryFactory $entity_query,
+    EntityTypeManagerInterface $entity_type_manager,
+    ActivityFactory $activity_factory,
+    GroupMuteNotify $group_mute_notify
+  ) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition, $entity_query, $entity_type_manager, $activity_factory);
```
Comment (navneet0693) [inline modules/custom/activity_basics/src/Plugin/ActivityContext/ContentInMyGroupActivityContext.php:60]: @Ressinel GroupMuteNotify parameter is missing 🤔

File: modules/custom/activity_basics/src/Plugin/ActivityContext/ContentInMyGroupActivityContext.php [hunk 14]
```diff
@@ -18,6 +23,60 @@
  */
 class ContentInMyGroupActivityContext extends ActivityContextBase {
 
+  /**
+   * The group mute notifications.
+   *
+   * @var \Drupal\social_group\GroupMuteNotify
+   */
+  protected $groupMuteNotify;
```
Comment (navneet0693) [inline modules/custom/activity_basics/src/Plugin/ActivityContext/ContentInMyGroupActivityContext.php:31]: protected GroupMuteNotify $groupMuteNotify;

File: modules/custom/activity_send/modules/activity_send_email/src/Plugin/QueueWorker/ActivitySendEmailWorker.php [hunk 15]
```diff
@@ -78,6 +80,13 @@ class ActivitySendEmailWorker extends ActivitySendWorkerBase implements Containe
    */
   protected $languageManager;
 
+  /**
+   * The group mute notifications.
+   *
+   * @var \Drupal\social_group\GroupMuteNotify
+   */
+  protected $groupMuteNotify;
```
Comment (navneet0693) [inline modules/custom/activity_send/modules/activity_send_email/src/Plugin/QueueWorker/ActivitySendEmailWorker.php:88]: `protected GroupMuteNotify $groupMuteNotify;`

File: modules/social_features/social_group/social_group.module [hunk 16]
```diff
@@ -2021,6 +2028,22 @@ function social_group_get_allowed_visibility_options_per_group_type($group_type_
...
```
Comment (navneet0693) [inline modules/social_features/social_group/social_group.module:2034]: Please update this documentation with a better explanation of what to remove.

File: modules/social_features/social_group/social_group.module [hunk 17]
```diff
@@ -2021,6 +2028,22 @@ function social_group_get_allowed_visibility_options_per_group_type($group_type_
...
+    // show unflagged groups.
+    // @see https://www.drupal.org/project/flag/issues/2841067#comment-14135283
```
Comment (navneet0693) [inline modules/social_features/social_group/social_group.module:2036]: Maybe file a separate issue for Groups or use https://www.drupal.org/project/group/issues/2952802

File: modules/social_features/social_group/src/GroupMuteNotify.php [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+  public function groupNotifyIsMuted(GroupInterface $group, AccountInterface $account): bool {
+    $flaggings = $this->flagService->getAllEntityFlaggings($group, $account);
+
+    return !empty($flaggings);
+  }
+
+  /**
+   * Get group by content.
+   *
+   * @param \Drupal\Core\Entity\EntityInterface $entity
```
Comment (navneet0693) [inline modules/social_features/social_group/src/GroupMuteNotify.php:59]: `* @param \Drupal\Core\Entity\EntityInterface|null $entity`

**Comment by navneet0693** [comment]
On an already existing installation of 10.2.x, I checkout to this branch and ran `drush updb`

As Chris Hall, I muted Official Group Educational Assistance Association and I logged out.

As Susan Williams, I logged in.

1. I mentioned Chris Hall in the comments.

Screenshot 2021-06-18 at 6 05 46 PM

2. I commented on topic/events/post created by Chris Hall.

Screenshot 2021-06-18 at 6 05 40 PM

3.  I liked post/event/topic by Chris Hall.

Screenshot 2021-06-18 at 6 05 12 PM

As Chris Hall, I received both notifications and emails about the above activities.

**Comment by navneet0693** [comment]
I enabled,  Welcome message module,

AS Chris Hall, I created a "Flexible group" - Public - Open to join and also enabled welcome message. I logged out.

AS Susan Williams, I logged in and I joined the above group created by Chris hall. 

I received a private message, and I logout.

As Chris Hall, I checked my inbox, no private message was created.

**Needed:** 
Private message should have been create like:

Screenshot 2021-06-18 at 7 28 56 PM

File: modules/custom/activity_basics/src/Plugin/ActivityContext/ContentInMyGroupActivityContext.php (hunk 14 above)
Comment (Ressinel) [reply modules/custom/activity_basics/src/Plugin/ActivityContext/ContentInMyGroupActivityContext.php]: Updated.

File: modules/custom/activity_send/modules/activity_send_email/src/Plugin/QueueWorker/ActivitySendEmailWorker.php (hunk 15 above)
Comment (Ressinel) [reply modules/custom/activity_send/modules/activity_send_email/src/Plugin/QueueWorker/ActivitySendEmailWorker.php]: Updated.

File: modules/social_features/social_group/social_group.module (hunk 16 above)
Comment (Ressinel) [reply modules/social_features/social_group/social_group.module]: Updated.

File: modules/social_features/social_group/social_group.module (hunk 17 above)
Comment (Ressinel) [reply modules/social_features/social_group/social_group.module]: Created new issue for Groups.

**Comment by Ressinel** [comment]
@navneet0693 

I have covered the cases you provided above. 

Regarding the welcome message I left as it was before since it is a correct scenario when a user doesn't receive a welcome message if group notifications are muted.

**Comment by navneet0693** [comment]
@Ressinel I see that Tugboat is failing because of unmet dependencies in view.view.group config. Shall we add dependency of better_exposed_filters on social_group module?

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2357-oateu14g0l3ipmdvyd5sgb2hsuacaee9.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60c86f723e5c64c1e2ad8cb3

**Comment by Ressinel** [comment]
@navneet0693 I have added dependency on `better_exposed_filters` in `social_group` module and looks like **Tugbot** is not failing now.

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2357-oateu14g0l3ipmdvyd5sgb2hsuacaee9.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60c86f723e5c64c1e2ad8cb3

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2357-oateu14g0l3ipmdvyd5sgb2hsuacaee9.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60c86f723e5c64c1e2ad8cb3

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2357-oateu14g0l3ipmdvyd5sgb2hsuacaee9.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60c86f723e5c64c1e2ad8cb3

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2357-oateu14g0l3ipmdvyd5sgb2hsuacaee9.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60c86f723e5c64c1e2ad8cb3

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2357-oateu14g0l3ipmdvyd5sgb2hsuacaee9.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60c86f723e5c64c1e2ad8cb3

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2357-oateu14g0l3ipmdvyd5sgb2hsuacaee9.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60c86f723e5c64c1e2ad8cb3

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2357-oateu14g0l3ipmdvyd5sgb2hsuacaee9.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60c86f723e5c64c1e2ad8cb3

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2357-oateu14g0l3ipmdvyd5sgb2hsuacaee9.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60c86f723e5c64c1e2ad8cb3

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2357-oateu14g0l3ipmdvyd5sgb2hsuacaee9.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60c86f723e5c64c1e2ad8cb3

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2357-oateu14g0l3ipmdvyd5sgb2hsuacaee9.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60c86f723e5c64c1e2ad8cb3

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2357-oateu14g0l3ipmdvyd5sgb2hsuacaee9.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60c86f723e5c64c1e2ad8cb3

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2357-oateu14g0l3ipmdvyd5sgb2hsuacaee9.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60c86f723e5c64c1e2ad8cb3

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2357-oateu14g0l3ipmdvyd5sgb2hsuacaee9.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60c86f723e5c64c1e2ad8cb3

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2357-oateu14g0l3ipmdvyd5sgb2hsuacaee9.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60c86f723e5c64c1e2ad8cb3

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2357-oateu14g0l3ipmdvyd5sgb2hsuacaee9.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60c86f723e5c64c1e2ad8cb3

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2357-oateu14g0l3ipmdvyd5sgb2hsuacaee9.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60c86f723e5c64c1e2ad8cb3

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2357-oateu14g0l3ipmdvyd5sgb2hsuacaee9.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60c86f723e5c64c1e2ad8cb3

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2357-oateu14g0l3ipmdvyd5sgb2hsuacaee9.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60c86f723e5c64c1e2ad8cb3

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2357-oateu14g0l3ipmdvyd5sgb2hsuacaee9.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60c86f723e5c64c1e2ad8cb3

File: modules/social_features/social_group/social_group.install [hunk 19]
```diff
@@ -696,3 +699,78 @@ function social_group_update_10302() {
   // Output logged messages to related channel of update execution.
   return $updateHelper->logger()->output();
 }
+
+/**
+ * Import flag "Mute Group Notifications" configurations.
+ */
+function social_group_update_10303() {
+  $configs = [
+    'install' => [
```
Comment (ronaldtebrake) [inline modules/social_features/social_group/social_group.install:708]: We can never use install folders in an update hook, always use a static representation of this. 
e.g. an array or a `config/static`

File: modules/social_features/social_group/social_group.install [hunk 20]
```diff
@@ -696,3 +699,78 @@ function social_group_update_10302() {
...
+  // Output logged messages to related channel of update execution.
+  return $updateHelper->logger()->output();
+}
+
+/**
+ * Import exposed block filter configurations of User Groups view.
+ */
+function social_group_update_10306() {
+  $configs = [
+    'optional' => [
```
Comment (ronaldtebrake) [inline modules/social_features/social_group/social_group.install:756]: Also counts for `config/optional` which is also not a static representation, it can change over time.
Imagine users coming from 10.0.0 and going to 11.0.0 in the future, the content of these yml files might have changed and could create issues with dependencies etc.

File: modules/social_features/social_group/social_group.module [hunk 21]
```diff
@@ -246,6 +249,15 @@ function social_group_preprocess_group(array &$variables) {
 
   $account = \Drupal::currentUser();
 
+  // Prevent access to mute group notifications if a user is not a group member
+  // and show that link for specific regions only.
+  $allowed_regions = [
```
Comment (ronaldtebrake) [inline modules/social_features/social_group/social_group.module:254]: Why is this necessary? 🤔 
Thats why we have our theme and view modes right? they dictate if a flag is rendered on a page in a view mode.

Imagine someone having a subtheme without these regions, or someone updating the config for a view mode and wanting to add it to a different region?

File: tests/behat/features/capabilities/group/group-mute-notify.feature [hunk 22]
```diff
@@ -0,0 +1,142 @@
+@api @notifications @group @DS-7598 @stability @stability-3 @group-mute-group-notifications
+Feature: Mute/Unmute group notifications
```
Comment (ronaldtebrake) [inline tests/behat/features/capabilities/group/group-mute-notify.feature:2]: Well done on writing this test!!!! 💚

File: modules/social_features/social_group/social_group.install (hunk 19 above)
Comment (Ressinel) [reply modules/social_features/social_group/social_group.install:708]: So, what is the best way to add new configurations now? By using update helper?

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2357-oateu14g0l3ipmdvyd5sgb2hsuacaee9.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60c86f723e5c64c1e2ad8cb3

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2357-oateu14g0l3ipmdvyd5sgb2hsuacaee9.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60c86f723e5c64c1e2ad8cb3

File: modules/social_features/social_group/social_group.install (hunk 19 above)
Comment (Ressinel) [reply modules/social_features/social_group/social_group.install]: I've changed it. Check it, please.

File: modules/social_features/social_group/social_group.install (hunk 20 above)
Comment (Ressinel) [reply modules/social_features/social_group/social_group.install]: I've changed it as well. Check it, please.

File: modules/social_features/social_group/social_group.module (hunk 21 above)
Comment (Ressinel) [reply modules/social_features/social_group/social_group.module]: You're right, I've moved it into view display settings.

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2357-oateu14g0l3ipmdvyd5sgb2hsuacaee9.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60c86f723e5c64c1e2ad8cb3

**Review by navneet0693 — APPROVED** [review]

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2357-oateu14g0l3ipmdvyd5sgb2hsuacaee9.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60c86f723e5c64c1e2ad8cb3
